agent/consul/intention_endpoint.go

```diff
@@ -171,336 +317,181 @@ func (s *Intention) prepareApplyUpdateLegacy(
-			}
+	upsertEntry.UpsertSourceByName(sn, args.Intention.ToSourceIntention(false))
 
-			found := false
-			for i, src := range upsertEntry.Sources {
-				if src.SourceServiceName() == sn {
-					upsertEntry.Sources[i] = args.Intention.ToSourceIntention(false)
-					found = true
-					break
-				}
-			}
-			if !found {
-				upsertEntry.Sources = append(upsertEntry.Sources, args.Intention.ToSourceIntention(false))
-			}
-		}
+	return structs.ConfigEntryUpsertCAS, upsertEntry, nil
+}
 
-	case structs.IntentionOpDelete:
-		// There are two ways to get this request:
-		//
-		// 1) legacy: the ID field is populated
-		// 2) config-entry: the ID field is NOT populated
+func (s *Intention) computeApplyChangesLegacyDelete(
+	accessorID string,
+	authz acl.Authorizer,
+	entMeta *structs.EnterpriseMeta,
+	args *structs.IntentionRequest,
+) (structs.ConfigEntryOp, *structs.ServiceIntentionsConfigEntry, error) {
+	_, prevEntry, ixn, err := s.srv.fsm.State().IntentionGet(nil, args.Intention.ID)
+	if err != nil {
+		return "", nil, fmt.Errorf("Intention lookup failed: %v", err)
+	}
+	if ixn == nil || prevEntry == nil {
+		return "", nil, fmt.Errorf("Cannot delete non-existent intention: '%s'", args.Intention.ID)
+	}
 
-		if args.Intention.ID == "" {
-			// config-entry style: no LegacyID
-			legacyWrite = false
+	if err := checkLegacyIntentionApplyAllowed(prevEntry); err != nil {
+		return "", nil, err
+	}
 
-			args.Intention.DefaultNamespaces(&entMeta)
+	// Even though the eventual config entry RPC will do an authz check and
+	// validation, if we do them here too we can generate error messages that
+	// make more sense for legacy edits.
+	if !ixn.CanWrite(authz) {
+		// todo(kit) Migrate intention access denial logging over to audit logging when we implement it
+		s.logger.Warn("Deletion operation on intention denied due to ACLs", "intention", args.Intention.ID, "accessorID", accessorID)
+		return "", nil, acl.ErrPermissionDenied
+	}
 
-			prevEntry, err = s.getServiceIntentionsConfigEntry(args.Intention.DestinationName, args.Intention.DestinationEnterpriseMeta())
-			if err != nil {
-				return err
-			}
+	upsertEntry := prevEntry.Clone()
 
-			// NOTE: validation errors may be misleading!
-			noop = true
-			if prevEntry != nil {
-				sn := args.Intention.SourceServiceName()
-
-				upsertEntry = prevEntry.Clone()
-				for i, src := range upsertEntry.Sources {
-					if src.SourceServiceName() == sn {
-						// Delete slice element: https://github.com/golang/go/wiki/SliceTricks#delete
-						//    a = append(a[:i], a[i+1:]...)
-						upsertEntry.Sources = append(upsertEntry.Sources[:i], upsertEntry.Sources[i+1:]...)
-
-						if len(upsertEntry.Sources) == 0 {
-							upsertEntry.Sources = nil
-						}
-						noop = false
-						break
-					}
-				}
-			}
+	deleted := upsertEntry.DeleteSourceByLegacyID(args.Intention.ID)
+	if !deleted {
+		return "", nil, fmt.Errorf("Cannot delete non-existent intention: '%s'", args.Intention.ID)
```

Comment:
Hmm explaining why this function failed from the outside seems a bit presumptive. It's fine now, but there could be other sources of failure down the line. Wondering if this would benefit from returning `delete, found` or maybe have a more generic error message.

Thoughts?

Reply:
So in this particular method `computeApplyChangesLegacyDelete` I was trying extra hard to keep the same weird semantics that the legacy intention delete-by-ID had.

In the newer delete-by-source-and-dest version, deletes are idempotent and if you delete something that doesn't exist it does nothing and returns :+1: